src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+
+    if (nbytes == 0)                    return 0;
+    uint32 bytes_to_write = nbytes;
+    uint32 bytes_remaining = F_PART_DATA_SIZE - position->part_offset;
+
+    //while there are enough bytes to be written to fill a f_part
+    while (bytes_to_write > bytes_remaining) {
+        memcpy(position->open_part->data + position->part_offset, buffer, bytes_remaining);
+        position->file_offset           += bytes_remaining;
+        buffer                          += bytes_remaining;
+        bytes_to_write                  -= bytes_remaining;
+        position->part_offset           = 0;
+        if (position->open_part->next == NULL) {
+            struct f_part *part;
+            newpart_get(&part);
+            part->file                  = o;
+            part->next                  = NULL;
+            part->prev                  = position->open_part;
+            position->open_part->next   = part;
+        }
+
+        position->open_part             = position->open_part->next;
+        bytes_remaining                 = F_PART_DATA_SIZE - position->part_offset;
+    }
+    //bytes_to_write < bytes_remaining
+    memcpy(position->open_part->data, buffer, bytes_to_write);
+    position->part_offset               += bytes_to_write;
+    position->file_offset               += bytes_to_write;
+    if (o->size < position->file_offset) {
+        o->size                         = position->file_offset;
+    }
+    return nbytes;
 }
 
-uint32 file_insert(struct fsobj *o, struct FS *filesys)
+struct fsobj *file_find(char *path)
 {
-    assert(filesys && filesys->root && filesys->root->filedes);
-    if (filesys->root == o) return OS_FS_SUCCESS;
+    assert(path);
+    //paths should always begin with '/' dir names do not
+    if (path[0] != '/') return NULL;
+    path++;
+
+    // token is the current directory in path
+    char path_temp[OS_MAX_PATH_LEN * 2];
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
+    char *token = strtok(path_temp, delim);
+
+    if(!openfs || !openfs->root) return NULL;
+    struct fsobj *cur = openfs->root;
+    assert(cur && cur->name);
+    if (strcmp(token, cur->name) != 0) return NULL;
+
+    while (1) {
+        // iterate through linked list of children until ancestor is found
+        while (strcmp(token, cur->name) != 0) {
+            if (!cur->next) return NULL;
+            cur = cur->next;
+        }
+        token = strtok(NULL, delim);
+        if (token == NULL) return cur;
+        if (!cur->child) return NULL;
+        cur = cur->child;
+    }
+    PANIC("Unreachable Statement");
+    return NULL;
+}
 
-    insert(o, filesys->root);
+int32 file_create(char *path, permission_t permission)
+{
+    assert(path);
+
+    struct fsobj *o;
+    if (newfile_get(&o) != OS_FS_SUCCESS) return OS_FS_ERROR;
+
+    o->name = path_to_name(path);
+    o->type = FSOBJ_FILE;
+    o->size = 0;
+    o->permission = permission;
+
+    struct f_part *part;
+    newpart_get(&part);
+    assert(part);
```

Comment:
@zacharied I think if the violation of a constraint always means there is a bug, then asserts are fine. Otherwise, I agree if you.